### PR TITLE
Misc additions

### PR DESCRIPTION
**Pull request by abizjak**
## Purpose

Add support for CIS0 queries, and fix some bugs in serialization of CIS2 events.

## Changes

- Add CIS0 module
- Fix parsing of CIS2 events, add helper TryFrom implementation.

## Checklist

- [ ] My code follows the style of this project.
- [ ] The code compiles without warnings.
- [ ] I have performed a self-review of the changes.
- [ ] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: src/cis0.rs [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+    /// The standard is supported by using another contract address.
+    SupportBy(Vec<ContractAddress>),
+}
+
+impl SupportResult {
+    /// Return whether the result is [`Support`](Self::Support) or not.
+    pub fn is_support(&self) -> bool { matches!(self, &Self::Support) }
+}
+
+impl contracts_common::Serial for SupportResult {
```
Comment (limemloh) [inline src/cis0.rs:33]: You could derive this instead.

File: src/cis0.rs [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+                for addr in addrs {
+                    addr.serial(out)?;
+                }
+                Ok(())
+            }
+        }
+    }
+}
+
+impl contracts_common::Deserial for SupportResult {
```
Comment (limemloh) [inline src/cis0.rs:50]: This could be derived as well

File: src/cis0.rs [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+impl contracts_common::Serial for SupportsQueryResponse {
+    fn serial<W: contracts_common::Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        (self.results.len() as u16).serial(out)?;
+        for result in &self.results {
+            result.serial(out)?;
+        }
+        Ok(())
+    }
+}
```
Comment (limemloh) [inline src/cis0.rs:96]: These could be derived

File: src/cis0.rs [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+pub enum SupportsError {
+    #[error("The name of the contract is not valid and thus the contract does not support CIS0.")]
+    ContractNameInvalid,
+    #[error("Parameter size exceeds maximum allowed. Too many ids.")]
+    InvalidParameter,
+    #[error("Query error: {0}")]
+    QueryError(#[from] super::v2::QueryError),
+    #[error("Contract reject.")]
+    ContractReject,
+    #[error("No return. This is a V0 contract.")]
```
Comment (limemloh) [inline src/cis0.rs:172]: Maybe add that: Thus the contracts does not support CIS0

**Review by limemloh — APPROVED** [review]
Looks good.
Only minor comments

File: src/cis0.rs (hunk 1 above)
Comment (abizjak) [reply src/cis0.rs:33]: Well technically yes, but in a hacky way. The contracts-common derive works either in a crate where you have access to concordium_rust_sdk or concordium_std.

I can derive **if** I add the following

```
use crate as concordium_rust_sdk
```
to the module where I derive serialization. But this to me looks very very ugly and I would prefer not to do it.

File: src/cis0.rs (hunk 2 above)
Comment (abizjak) [reply src/cis0.rs:50]: As above

File: src/cis0.rs (hunk 3 above)
Comment (abizjak) [reply src/cis0.rs:96]: As above.

File: src/cis0.rs (hunk 4 above)
Comment (abizjak) [reply src/cis0.rs:172]: Makes sense.

**Pull request merged**